src/dewey/indexing.clj

```diff
@@ -1,72 +1,81 @@
-
-(defmethod mapping-type-of Keyword
-  [entity-type]
-  (case entity-type
-    :collection  collection-type
-    :data-object data-object-type))
-
-
-(defn- index-doc
-  [es mapping-type doc]
-  (es-doc/create es (cfg/es-index) mapping-type doc :id (str (:id doc))))
-
+   Throws:
+     This function can throw an exception if it can't connect to elasticsearch or iRODS. The
+     function can also throw one if the document is already indexed."
+  [es doc]
+  (s/request es {:url
+                 [(cfg/es-index) :_create (str (:id doc))]
+                 :method :put
+                 :headers {"Content-Type" "application/json"}
+                 :body doc}))
 
 (defn- update-doc
   "Scripted updates which are only compatible with Elasticsearch 5.x and greater."
   [es entity script params]
-  (rest/post es
-             (rest/record-update-url es
-                                     (cfg/es-index)
-                                     (mapping-type-of entity)
-                                     (str (entity/id entity)))
-             {:body {:script {:inline script :lang "painless" :params params}}}))
+  (s/request es {:url [(cfg/es-index) :_update (str (entity/id entity))]
+                 :method :post
+                 :headers {"Content-Type" "application/json"}
+                 :body {"script" {"source" script "lang" "painless" "params" params}}}))
+
+(defn- index-error
+  [e]
+  (let [resp (ex-data e)]    (if
+                              (= 404 (:status resp))
+                               false
+                               (do (log/info "Elasticsearch is not responding as expected.")
+                                   (throw e)))))
 
-(defn entity-indexed?
-  ([es entity]
-   ^{:doc "Determines whether or not an iRODS entity has been indexed.
+
+(defmulti
+  ^{:doc "Determines whether or not an iRODS entity has been indexed.
 
            Parameters:
              es     - the elasticsearch connection
              entity - the entity being checked
 
            Throws:
              This function can throw an exception if it can't connect to elasticsearch."}
-   (es-doc/present? es (cfg/es-index) (mapping-type-of entity) (str (entity/id entity))))
+  entity-indexed? (fn [_es entity] (cond
+                                     (string? entity) :string
+                                     (map? entity) :map)))
 
-  ([es entity-type entity-id]
-   ^{:doc "Determines whether or not an iRODS entity has been indexed.
 
-           Parameters:
-             es          - the elasticsearch connection
-             entity-type - :collection|:data-object
-             entity-id   - the UUID of the entity being checked
+(defmethod entity-indexed? :string
+  [es entity-id]
+  (try+
+   (s/request es {:url [(cfg/es-index) :_doc entity-id]
+                  :method :head})
+   true
+   (catch clojure.lang.ExceptionInfo e ;;qbits.spandex.ResponseException is wrapped in clojure.lang.ExceptionInfo
+     (index-error e))))
```

Comment:
This works the way it is, but I think that it might be possible to make it a little more succinct by using Slingshot's predicate feature. The way to do that would be to first define a predicate that determines whether or not an exception indicates that an entity wasn't found:

``` clojure
(defn not-found-exception?
  "Determines whether or not an exception indicates that an entity wasn't found.

     Parameters:
       ex - an exception that has been wrapped using clojure.core/ex-info"
  [ex]
  (= 404 (:status (ex-data ex))))
```
Once you have that, this method can become:

``` clojure
(defmethod entity-indexed? :string
  [es entity-id]
  (try+
   (s/request es {:url [(cfg/es-index) :_doc entity-id]
                  :method :head})
   true
   (catch not-found-exception? _ false)))
```

Of course, the same thing applies to the version of this multimethod for maps.

Reply:
I was hoping that we could do something like `(catch [:status 404] _ false)`, but that didn't work. 😞